.github/workflows/ci.yml

```diff
@@ -1,90 +1,98 @@
-                ports:
-                    - 3306:3306
-                options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=5
-        steps:
-            - uses: actions/checkout@v3
-
-            - name: Setup WooCommerce Monorepo
-              uses: ./.github/actions/setup-woocommerce-monorepo
-              with:
-                  build-filters: woocommerce
-
-            - name: Tool versions
-              run: |
-                  php --version
-                  composer --version
-
-            - name: Build Admin feature config
-              working-directory: plugins/woocommerce
-              run: pnpm run build:feature-config
-
-            - id: parseMatrix
-              name: Parse Matrix Variables
-              uses: actions/github-script@v6
-              with:
-                  script: |
-                      const parseWPVersion = require( './.github/workflows/scripts/parse-wp-version' );
-                      parseWPVersion( '${{ matrix.wp }}' ).then( ( version ) => {
-                        core.setOutput( 'wpVersion', version );
-                      } );
-
-            - name: Prepare Testing Environment
-              env:
-                  WP_ENV_CORE: ${{ steps.parseMatrix.outputs.wpVersion }}
-                  WP_ENV_PHP_VERSION: ${{ matrix.php }}
-              run: pnpm --filter=woocommerce env:test
-
-            - name: Run Tests
-              env:
-                  WP_ENV_CORE: ${{ steps.parseMatrix.outputs.wpVersion }}
-                  WP_ENV_PHP_VERSION: ${{ matrix.php }}
-              run: pnpm --filter=woocommerce test:unit:env --testsuite ${{ matrix.unittests }}
+    project-matrix:
+      # Since this is a monorepo, not every pull request or change is going to impact every project.
+      # Instead of running CI tasks on all projects indiscriminately, we use a script to detect
+      # which projects have changed and what kind of change occurred. This lets us build a
+      # matrix that we can use to run CI tasks only on the projects that need them.
+      name: 'Build Project Matrix'
+      runs-on: 'ubuntu-20.04'
+      outputs:
+        matrix: ${{ steps.project-matrix.outputs.matrix }}
+      steps:
+        - uses: 'actions/checkout@v3'
+          name: 'Checkout'
+          with:
+            fetch-depth: 0
+        - uses: './.github/actions/setup-woocommerce-monorepo'
+          name: 'Setup Monorepo'
+          with:
+            install: false
+        - uses: actions/github-script@v6
+          id: 'project-matrix'
+          name: 'Build Matrix'
+          with:
+            script: |
+              let baseRef = ${{ toJson( github.base_ref ) }};
+              if ( baseRef ) {
+                baseRef = 'origin/' + baseRef;
+              }
+              const buildCIMatrix = require( './.github/workflows/scripts/build-ci-matrix' );
+              core.setOutput( 'matrix', JSON.stringify( await buildCIMatrix( baseRef ) ) );
```

Comment:
Some logging would be useful here.

I introduced [a lint error in `@woocommerce/components`](https://github.com/woocommerce/woocommerce/pull/40691/commits/79422d1c070639564d199b50134ab2904f8b0775) and I do get a failure in JS tests for `woocommerce/client/admin`. I'm trying to debug and see if that's even related. `@woocommerce/components` tests weren't run. See [PR here](https://github.com/woocommerce/woocommerce/pull/40691)

Reply:
Some logging in the Evaluate Matrix step would be useful too, the error isn't too helpful

Reply:
Upon restarting, I do get the components to lint, but it passes 🤷🏽 https://github.com/woocommerce/woocommerce/actions/runs/6476659736/job/17588896320?pr=40691

Reply:
> Some logging would be useful here.

What kind of logging would you be interested in? The output is the matrix, but you can see what jobs were created and the like. The only information that I can think of that might be "hidden" is the intermediary steps leading up to the creation of the matrix. Namely, the files that were changed (and the projects they marked as changed) and the changes that were marked (along with the file that marked them)? This seems like it _could_ be interesting for being able to track how a project got added to the matrix. I'll take a look at adding that.

> Some logging in the Evaluate Matrix step would be useful too, the error isn't too helpful

I'm not really sure what value we could extract from this step. All it does is check for failed jobs so the error indicating that a job has failed is just telling you what you can already see from GitHub's interface on the job list. The `needs` context doesn't provide us information about _which_ jobs failed, but again, you can actually just see that. Once you click on the job you'll see what steps failed.

Reply:
>What kind of logging would you be interested in? The output is the matrix, but you can see what jobs were created and the like

I made a change in a particular file which should kick off certain checks, which didn't fire. My attempts to figure out what went wrong (at the matrix creation step? At the matrix interpretation step?) would have been aided by some logging. For instance, a print out of the matrix itself would have let me know that indeed the changes made were correctly picked up (or not) and given me a clue as to where the problem might be. 

I find it useful for tools like this to be overly descriptive. "Here's what I'm dong", "Here's what the result is", and "Here is what I'm going to do next".

Reply:
I added some pretty extensive logging, take a look and let me know what you think.